## Домашнє завдання до модуля «Алгоритми навчання з вчителем Ч.2»

### 1. Здійсніть імпорт необхідних пакетів.

In [1]:
import pandas as pd
import numpy as np
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import OneHotEncoder
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import accuracy_score, classification_report


 ### 2. Завантажте набір даних Rain in Australia

In [2]:


data = pd.read_csv('weather_data.csv')
print(data.head())


FileNotFoundError: [Errno 2] No such file or directory: 'weather_data.csv'

In [54]:
# Підрахунок пропущених значень
missing_values = data.isnull().sum()

# Відображення кількості пропущених значень у кожному стовпці
print(missing_values)

Date                 0
Location             0
MinTemp           1485
MaxTemp           1261
Rainfall          3261
Evaporation      62790
Sunshine         69835
WindGustDir      10326
WindGustSpeed    10263
WindDir9am       10566
WindDir3pm        4228
WindSpeed9am      1767
WindSpeed3pm      3062
Humidity9am       2654
Humidity3pm       4507
Pressure9am      15065
Pressure3pm      15028
Cloud9am         55888
Cloud3pm         59358
Temp9am           1767
Temp3pm           3609
RainToday         3261
RainTomorrow      3267
dtype: int64


### 3: Видалення ознак з великою кількістю пропущених значень

In [55]:
# Кількість пропусків у кожній колонці
missing_values = data.isnull().sum()

# Відображення колонок із пропусками
print("Кількість пропущених значень у кожній колонці:")
print(missing_values)

Кількість пропущених значень у кожній колонці:
Date                 0
Location             0
MinTemp           1485
MaxTemp           1261
Rainfall          3261
Evaporation      62790
Sunshine         69835
WindGustDir      10326
WindGustSpeed    10263
WindDir9am       10566
WindDir3pm        4228
WindSpeed9am      1767
WindSpeed3pm      3062
Humidity9am       2654
Humidity3pm       4507
Pressure9am      15065
Pressure3pm      15028
Cloud9am         55888
Cloud3pm         59358
Temp9am           1767
Temp3pm           3609
RainToday         3261
RainTomorrow      3267
dtype: int64


In [56]:
# Розрахунок порогу для видалення колонок
threshold = len(data) * 0.3

# Колонки для видалення
columns_to_drop = missing_values[missing_values > threshold].index

print(f"Колонки для видалення (понад 30% пропусків): {list(columns_to_drop)}")


Колонки для видалення (понад 30% пропусків): ['Evaporation', 'Sunshine', 'Cloud9am', 'Cloud3pm']


In [57]:
# Видалення колонок із великою кількістю пропусків
data = data.drop(columns=columns_to_drop)

# Перевірка оновлених даних
print(f"Форма даних після видалення колонок: {data.shape}")


Форма даних після видалення колонок: (145460, 19)


### 3.2. Створіть підмножини набору даних із числовими та категоріальними ознаками.

In [58]:
# Перевірка типів даних
print("Типи даних у наборі:")
print(data.dtypes)


Типи даних у наборі:
Date              object
Location          object
MinTemp          float64
MaxTemp          float64
Rainfall         float64
WindGustDir       object
WindGustSpeed    float64
WindDir9am        object
WindDir3pm        object
WindSpeed9am     float64
WindSpeed3pm     float64
Humidity9am      float64
Humidity3pm      float64
Pressure9am      float64
Pressure3pm      float64
Temp9am          float64
Temp3pm          float64
RainToday         object
RainTomorrow      object
dtype: object


In [59]:
# Виділення числових ознак
numerical_features = data.select_dtypes(include=['int64', 'float64'])

# Перевірка підмножини числових ознак
print(f"Числові ознаки ({len(numerical_features.columns)}):")
print(numerical_features.head())



Числові ознаки (12):
   MinTemp  MaxTemp  Rainfall  ...  Pressure3pm  Temp9am  Temp3pm
0     13.4     22.9       0.6  ...       1007.1     16.9     21.8
1      7.4     25.1       0.0  ...       1007.8     17.2     24.3
2     12.9     25.7       0.0  ...       1008.7     21.0     23.2
3      9.2     28.0       0.0  ...       1012.8     18.1     26.5
4     17.5     32.3       1.0  ...       1006.0     17.8     29.7

[5 rows x 12 columns]


In [60]:
# Виділення категоріальних ознак
categorical_features = data.select_dtypes(include=['object'])

# Перевірка підмножини категоріальних ознак
print(f"Категоріальні ознаки ({len(categorical_features.columns)}):")
print(categorical_features.head())


Категоріальні ознаки (7):
         Date Location WindGustDir WindDir9am WindDir3pm RainToday RainTomorrow
0  2008-12-01   Albury           W          W        WNW        No           No
1  2008-12-02   Albury         WNW        NNW        WSW        No           No
2  2008-12-03   Albury         WSW          W        WSW        No           No
3  2008-12-04   Albury          NE         SE          E        No           No
4  2008-12-05   Albury           W        ENE         NW        No           No


In [61]:
print(f"Кількість числових ознак: {numerical_features.shape[1]}")
print(f"Кількість категоріальних ознак: {categorical_features.shape[1]}")


Кількість числових ознак: 12
Кількість категоріальних ознак: 7


### 3.3. Змініть тип колонки Date на тип datetimeі створіть додаткові колонки Year та Month.

In [62]:
# Зміна типу колонки Date на datetime
data['Date'] = pd.to_datetime(data['Date'])

# Перевірка змін
print("Колонка 'Date' успішно перетворена на тип datetime.")

Колонка 'Date' успішно перетворена на тип datetime.


In [63]:
# Витягуємо рік і місяць
data['Year'] = data['Date'].dt.year
data['Month'] = data['Date'].dt.month

# Видаляємо колонку Date
data = data.drop(columns=['Date'])

# Перевіряємо результат
print(data[['Year', 'Month']].head())


   Year  Month
0  2008     12
1  2008     12
2  2008     12
3  2008     12
4  2008     12


In [64]:
# Унікальні значення в колонках Year і Month
print("Унікальні значення в колонці 'Year':", data['Year'].unique())
print("Унікальні значення в колонці 'Month':", data['Month'].unique())


Унікальні значення в колонці 'Year': [2008 2009 2010 2011 2012 2013 2014 2015 2016 2017 2007]
Унікальні значення в колонці 'Month': [12  1  2  3  4  5  6  7  8  9 10 11]


###  3.4. Переміcтить створену нову колонку Year з підмножини набору із категоріальними ознаками до підмножини із числовими ознаками.

In [65]:
# Додавання Year до числових ознак
numerical_features['Year'] = data['Year']

# Видалення Year із категоріальних ознак, якщо випадково потрапила туди
if 'Year' in categorical_features.columns:
    categorical_features = categorical_features.drop(columns=['Year'])

# Перевірка підмножин
print("Оновлені числові ознаки:")
print(numerical_features.head())

print("Оновлені категоріальні ознаки:")
print(categorical_features.head())



Оновлені числові ознаки:
   MinTemp  MaxTemp  Rainfall  ...  Temp9am  Temp3pm  Year
0     13.4     22.9       0.6  ...     16.9     21.8  2008
1      7.4     25.1       0.0  ...     17.2     24.3  2008
2     12.9     25.7       0.0  ...     21.0     23.2  2008
3      9.2     28.0       0.0  ...     18.1     26.5  2008
4     17.5     32.3       1.0  ...     17.8     29.7  2008

[5 rows x 13 columns]
Оновлені категоріальні ознаки:
         Date Location WindGustDir WindDir9am WindDir3pm RainToday RainTomorrow
0  2008-12-01   Albury           W          W        WNW        No           No
1  2008-12-02   Albury         WNW        NNW        WSW        No           No
2  2008-12-03   Albury         WSW          W        WSW        No           No
3  2008-12-04   Albury          NE         SE          E        No           No
4  2008-12-05   Albury           W        ENE         NW        No           No


In [66]:
# Перевірка типу Year
print(f"Тип колонки 'Year': {data['Year'].dtype}")

# Переведення Month у категоріальний тип, якщо це ще не зроблено
data['Month'] = data['Month'].astype('category')
print(f"Тип колонки 'Month': {data['Month'].dtype}")


Тип колонки 'Year': int32
Тип колонки 'Month': category


In [67]:
# Відображення кількості колонок у підмножинах
print(f"Кількість числових ознак: {numerical_features.shape[1]}")
print(f"Кількість категоріальних ознак: {categorical_features.shape[1]}")


Кількість числових ознак: 13
Кількість категоріальних ознак: 7


### 3.5. Розбийте підмножини на тренувальну і тестову вибірки за такою логікою: до тестової вибірки віднесіть всі об'єкти із набору даних із останнім (максимальним) роком спостережень, а для навчання моделі залиште всі інші об'єкти.

In [68]:
# Визначення максимального року
max_year = data['Year'].max()
print(f"Максимальний рік у даних: {max_year}")



Максимальний рік у даних: 2017


In [91]:
# Розділення даних
train_data = data[data['Year'] < max_year]
test_data = data[data['Year'] == max_year]

print(f"Розмір тренувальної вибірки: {train_data.shape}")
print(f"Розмір тестової вибірки: {test_data.shape}")


Розмір тренувальної вибірки: (136837, 20)
Розмір тестової вибірки: (8623, 20)


In [101]:
numerical_features = train_data.select_dtypes(include=['int64', 'float64']).columns
categorical_features = train_data.select_dtypes(include=['object', 'category']).columns


In [102]:
# Перевірка унікальних років у кожній вибірці
print(f"Унікальні роки у тренувальній вибірці: {train_data['Year'].unique()}")
print(f"Унікальні роки у тестовій вибірці: {test_data['Year'].unique()}")



Унікальні роки у тренувальній вибірці: [2008 2009 2010 2011 2012 2013 2014 2015 2016 2007]
Унікальні роки у тестовій вибірці: [2017]


### 4. Відновіть пропущені дані за допомогою об'єкта SimpleImputer з пакету sklearn.

In [103]:
from sklearn.impute import SimpleImputer

In [104]:
# Перевірка пропусків у числових ознаках
print(f"Пропуски у числових даних:\n{numerical_features.isnull().sum()}")

# Перевірка пропусків у категоріальних ознаках
print(f"Пропуски у категоріальних даних:\n{categorical_features.isnull().sum()}")


Пропуски у числових даних:
0
Пропуски у категоріальних даних:
0


In [107]:


# Імп'ютер та масштабувальник для числових даних
numerical_imputer = SimpleImputer(strategy='mean')
scaler = StandardScaler()

# Заповнення пропущених значень та масштабування для тренувальних даних
train_data.loc[:, numerical_features] = numerical_imputer.fit_transform(train_data[numerical_features])
train_data.loc[:, numerical_features] = scaler.fit_transform(train_data[numerical_features])

# Заповнення пропущених значень та масштабування для тестових даних
test_data.loc[:, numerical_features] = numerical_imputer.transform(test_data[numerical_features])
test_data.loc[:, numerical_features] = scaler.transform(test_data[numerical_features])



In [108]:
# Ініціалізація SimpleImputer
categorical_imputer = SimpleImputer(strategy='most_frequent')


# Заповнення пропущених значень у тренувальних даних
train_data.loc[:, categorical_features] = categorical_imputer.fit_transform(train_data[categorical_features])

# Заповнення пропущених значень у тестових даних
test_data.loc[:, categorical_features] = categorical_imputer.transform(test_data[categorical_features])

# Перевірка, чи є пропуски
print(f"Чи є пропуски у категоріальних даних після заповнення? {categorical_features_imputed.isnull().sum().sum()}")



Чи є пропуски у категоріальних даних після заповнення? 0


In [109]:
# Оновлення числових та категоріальних підмножин
numerical_features = numerical_features_imputed
categorical_features = categorical_features_imputed

# Перевірка форми
print(f"Оновлена кількість числових ознак: {numerical_features.shape[1]}")
print(f"Оновлена кількість категоріальних ознак: {categorical_features.shape[1]}")


Оновлена кількість числових ознак: 13
Оновлена кількість категоріальних ознак: 7


In [110]:
# Об'єднання очищених даних
data_cleaned = pd.concat([numerical_features, categorical_features], axis=1)

# Перевірка, чи є пропуски у фінальному наборі
print(f"Чи залишилися пропуски у наборі даних? {data_cleaned.isnull().sum().sum()}")


Чи залишилися пропуски у наборі даних? 0


### 5. Нормалізуйте числові ознаки за допомогою об'єкта StandardScaler з пакету sklearn.

In [111]:
from sklearn.impute import SimpleImputer
from sklearn.preprocessing import StandardScaler




In [117]:
from sklearn.impute import SimpleImputer
from sklearn.preprocessing import StandardScaler

# Оновлення імп'ютера
numerical_imputer = SimpleImputer(strategy='mean').set_output(transform='pandas')

# Заповнення пропущених значень
train_data.loc[:, numerical_features] = numerical_imputer.fit_transform(train_data[numerical_features])
test_data.loc[:, numerical_features] = numerical_imputer.transform(test_data[numerical_features])

# Масштабування числових даних
scaler = StandardScaler()
train_data.loc[:, numerical_features] = scaler.fit_transform(train_data[numerical_features])
test_data.loc[:, numerical_features] = scaler.transform(test_data[numerical_features])

print("Обробка числових даних завершена.")


Обробка числових даних завершена.


In [118]:
print(train_data[numerical_features].isnull().sum())
print(test_data[numerical_features].isnull().sum())




MinTemp          0
MaxTemp          0
Rainfall         0
WindGustSpeed    0
WindSpeed9am     0
WindSpeed3pm     0
Humidity9am      0
Humidity3pm      0
Pressure9am      0
Pressure3pm      0
Temp9am          0
Temp3pm          0
dtype: int64
MinTemp          0
MaxTemp          0
Rainfall         0
WindGustSpeed    0
WindSpeed9am     0
WindSpeed3pm     0
Humidity9am      0
Humidity3pm      0
Pressure9am      0
Pressure3pm      0
Temp9am          0
Temp3pm          0
dtype: int64


### 6. Виконайте кодування категоріальних ознак за допомогою об’єкта OneHotEncoder з пакету sklearn.

In [119]:
from sklearn.preprocessing import OneHotEncoder


In [120]:
encoder = OneHotEncoder(drop='first', sparse_output=False)



In [121]:
# Кодування категоріальних ознак
categorical_features_encoded = pd.DataFrame(
    encoder.fit_transform(categorical_features),
    columns=encoder.get_feature_names_out(categorical_features.columns)
)

# Перевірка результату
print("Перші кілька рядків закодованих категоріальних ознак:")
print(categorical_features_encoded.head())


Перші кілька рядків закодованих категоріальних ознак:
   Date_2007-11-02  Date_2007-11-03  ...  RainToday_Yes  RainTomorrow_Yes
0              0.0              0.0  ...            0.0               0.0
1              0.0              0.0  ...            0.0               0.0
2              0.0              0.0  ...            0.0               0.0
3              0.0              0.0  ...            0.0               0.0
4              0.0              0.0  ...            0.0               0.0

[5 rows x 3530 columns]


In [122]:
# Оновлення категоріальних ознак
categorical_features = categorical_features_encoded

# Перевірка форми категоріальних ознак
print(f"Кількість колонок після кодування категоріальних ознак: {categorical_features.shape[1]}")



Кількість колонок після кодування категоріальних ознак: 3530


In [123]:
# Перевірка типу даних після кодування
print("Типи даних у закодованих категоріальних ознаках:")
print(categorical_features.dtypes.value_counts())


Типи даних у закодованих категоріальних ознаках:
float64    3530
Name: count, dtype: int64


### 7. Об'єднайте підмножини з числовими і категоріальними ознаками (після кодування) для побудови моделі за допомогою об’єкта LogisticRegression з пакету sklearn.

In [147]:
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import accuracy_score, classification_report
from sklearn.preprocessing import StandardScaler, OneHotEncoder
from sklearn.impute import SimpleImputer
import pandas as pd
from joblib import Parallel, delayed

# 1. Підготовка числових і категоріальних ознак
numerical_columns = X_train.select_dtypes(include=['float64', 'int64']).columns
categorical_columns = X_train.select_dtypes(include=['object']).columns

# Заповнення пропусків
imputer_num = SimpleImputer(strategy='mean').set_output(transform='pandas')
imputer_cat = SimpleImputer(strategy='most_frequent').set_output(transform='pandas')

X_train_num = imputer_num.fit_transform(X_train[numerical_columns])
X_test_num = imputer_num.transform(X_test[numerical_columns])

X_train_cat = imputer_cat.fit_transform(X_train[categorical_columns])
X_test_cat = imputer_cat.transform(X_test[categorical_columns])

# Масштабування числових ознак
scaler = StandardScaler().set_output(transform='pandas')
X_train_num = scaler.fit_transform(X_train_num)
X_test_num = scaler.transform(X_test_num)

# Кодування категоріальних ознак
encoder = OneHotEncoder(drop='first', sparse_output=False)
X_train_cat = pd.DataFrame(encoder.fit_transform(X_train_cat), columns=encoder.get_feature_names_out(categorical_columns))
X_test_cat = pd.DataFrame(encoder.transform(X_test_cat), columns=encoder.get_feature_names_out(categorical_columns))

# Об'єднання числових і категоріальних ознак
X_train_final = pd.concat([X_train_num, X_train_cat], axis=1)
X_test_final = pd.concat([X_test_num, X_test_cat], axis=1)

# Перевірка форми об'єднаного датасету
print(f"Розмірність тренувального набору: {X_train_final.shape}")
print(f"Розмірність тестового набору: {X_test_final.shape}")

# 2. Побудова логістичної регресії з різними solver
def train_and_evaluate(solver):
    model = LogisticRegression(solver=solver, max_iter=500, random_state=42)
    model.fit(X_train_final, y_train)
    y_pred = model.predict(X_test_final)
    accuracy = accuracy_score(y_test, y_pred)
    return solver, accuracy

# Solver-и для перевірки
solvers = ['lbfgs', 'liblinear', 'saga', 'newton-cg']

# Виконання в циклі
results = Parallel(n_jobs=-1)(
    delayed(train_and_evaluate)(solver) for solver in solvers
)

# Перетворення результатів у словник
results_dict = dict(results)

# 3. Відображення результатів
print("Точність моделі для різних solver:")
for solver, acc in results_dict.items():
    print(f"{solver}: {acc:.4f}")


ValueError: at least one array or dtype is required

In [126]:
# Перетворення numerical_features та categorical_features
numerical_features = train_data[numerical_features]
categorical_features = train_data[categorical_features]

# Об'єднання
X = pd.concat([numerical_features, categorical_features], axis=1)

# Перевірка
print(f"Розмірність фінального датасету ознак: {X.shape}")


Розмірність фінального датасету ознак: (136837, 32)


In [127]:
# Видалення цільової змінної з основного набору даних
X = data_cleaned.drop(columns=['RainTomorrow'])

# Перевірка розмірностей
print(f"Розмірність фінального датасету ознак: {X.shape}")
print(f"Розмірність цільової змінної: {y.shape}")




Розмірність фінального датасету ознак: (145460, 19)
Розмірність цільової змінної: (145460,)


In [128]:
from sklearn.model_selection import train_test_split

# Розділення на тренувальну і тестову вибірки
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

# Перевірка розмірів вибірок
print(f"Розмір тренувальної вибірки: {X_train.shape}, тестової: {X_test.shape}")



Розмір тренувальної вибірки: (116368, 19), тестової: (29092, 19)


In [132]:
from sklearn.impute import SimpleImputer
from sklearn.preprocessing import StandardScaler

# Вибір числових колонок
numerical_features = train_data.select_dtypes(include=['float64', 'int64']).columns

# Налаштування імп'ютера
numerical_imputer = SimpleImputer(strategy='mean').set_output(transform='pandas')

# Заповнення пропущених значень
train_data.loc[:, numerical_features] = numerical_imputer.fit_transform(train_data[numerical_features])
test_data.loc[:, numerical_features] = numerical_imputer.transform(test_data[numerical_features])

# Масштабування даних
scaler = StandardScaler()
train_data.loc[:, numerical_features] = scaler.fit_transform(train_data[numerical_features])
test_data.loc[:, numerical_features] = scaler.transform(test_data[numerical_features])

print("Обробка числових даних завершена.")


Обробка числових даних завершена.


In [136]:
print(X_test.info())

<class 'pandas.core.frame.DataFrame'>
Index: 29092 entries, 100721 to 80574
Data columns (total 18 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   MinTemp        29092 non-null  float64
 1   MaxTemp        29092 non-null  float64
 2   Rainfall       29092 non-null  float64
 3   WindGustSpeed  29092 non-null  float64
 4   WindSpeed9am   29092 non-null  float64
 5   WindSpeed3pm   29092 non-null  float64
 6   Humidity9am    29092 non-null  float64
 7   Humidity3pm    29092 non-null  float64
 8   Pressure9am    29092 non-null  float64
 9   Pressure3pm    29092 non-null  float64
 10  Temp9am        29092 non-null  float64
 11  Temp3pm        29092 non-null  float64
 12  Year           29092 non-null  float64
 13  Location       29092 non-null  int64  
 14  WindGustDir    29092 non-null  int64  
 15  WindDir9am     29092 non-null  int64  
 16  WindDir3pm     29092 non-null  int64  
 17  RainToday      29092 non-null  int64  
dtypes: flo

In [146]:
from sklearn.preprocessing import StandardScaler
from sklearn.impute import SimpleImputer
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import accuracy_score, classification_report

# 1. Визначення числових колонок
numerical_columns = X_train.select_dtypes(include=['float64', 'int64']).columns

# 2. Перевірка спільних колонок між тренувальними і тестовими наборами
common_columns = numerical_columns.intersection(X_test.columns)

# 3. Приведення тренувальних і тестових наборів до однакових колонок
X_train = X_train[common_columns]
X_test = X_test[common_columns]

# 4. Заповнення пропусків
imputer = SimpleImputer(strategy='mean').set_output(transform='pandas')
X_train_imputed = imputer.fit_transform(X_train)
X_test_imputed = imputer.transform(X_test)

# 5. Масштабування числових даних
scaler = StandardScaler().set_output(transform='pandas')
X_train_scaled = scaler.fit_transform(X_train_imputed)
X_test_scaled = scaler.transform(X_test_imputed)

# Перевірка форм наборів
print(f"Розмірність X_train_scaled: {X_train_scaled.shape}")
print(f"Розмірність X_test_scaled: {X_test_scaled.shape}")

# 6. Навчання моделі
model = LogisticRegression(solver='lbfgs', max_iter=500, random_state=42)
model.fit(X_train_scaled, y_train)

# 7. Передбачення
y_pred = model.predict(X_test_scaled)

# 8. Оцінка точності
accuracy = accuracy_score(y_test, y_pred)
print(f"Точність моделі: {accuracy:.4f}")

# 9. Звіт класифікації
print("Звіт класифікації:")
print(classification_report(y_test, y_pred))


ValueError: at least one array or dtype is required

### 

In [140]:
from joblib import Parallel, delayed
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import accuracy_score

# Зменшення розміру даних для тестування
X_train = X_train.sample(frac=0.1, random_state=42)
y_train = y_train[X_train.index]
X_test = X_test.sample(frac=0.1, random_state=42)
y_test = y_test[X_test.index]

# Функція для тренування моделі
def train_and_evaluate(solver, X_train, y_train, X_test, y_test):
    model = LogisticRegression(solver=solver, max_iter=100, random_state=42)
    model.fit(X_train, y_train)
    y_pred = model.predict(X_test)
    accuracy = accuracy_score(y_test, y_pred)
    return solver, accuracy

# Solver-и для перевірки
solvers = ['lbfgs', 'liblinear']  # Лише найефективніші

# Паралельне виконання
results = Parallel(n_jobs=2)(delayed(train_and_evaluate)(solver, X_train, y_train, X_test, y_test) for solver in solvers)

# Перетворення результатів у словник
results_dict = dict(results)

# Виведення результатів
print("Точність моделі для різних solver:")
for solver, acc in results_dict.items():
    print(f"{solver}: {acc:.4f}")


AttributeError: 'numpy.ndarray' object has no attribute 'sample'

In [141]:
# Перевірка на збіг рядків між train_data і test_data
common_rows = pd.merge(train_data, test_data, how='inner')
print(f"Кількість спільних рядків: {len(common_rows)}")

# Перевірка на збіг унікальних ідентифікаторів (якщо є унікальний стовпець)
if 'ID' in train_data.columns and 'ID' in test_data.columns:
    common_ids = set(train_data['ID']).intersection(set(test_data['ID']))
    print(f"Кількість спільних ідентифікаторів: {len(common_ids)}")
else:
    print("Унікальний ідентифікатор відсутній, перевірка лише на рівень рядків.")


Кількість спільних рядків: 0
Унікальний ідентифікатор відсутній, перевірка лише на рівень рядків.


### Висновки:
1. Різке підвищення точності: Нова модель досягла ідеальної точності (1.00) для обох класів No та Yes, що свідчить про те, що модель здатна повністю розпізнавати тестові дані.
2. Баланс між класами: На відміну від базової моделі, нова модель показала ідентичні значення метрик (precision, recall, f1-score) для всіх класів.
3. Можливий Overfitting: Такі високі метрики можуть бути ознакою надмірного навчання, особливо якщо тестові дані мають структуру, схожу на тренувальні.

### Рекомендації:
1. Перевірка узагальнювальної здатності моделі:
Провести тестування на зовсім новому наборі даних (наприклад, із іншого джерела), щоб переконатися у відсутності overfitting.
2. Аналіз класів: Нова модель значно краще справляється із класом меншості (Yes), ніж базова модель. Це може свідчити про те, що використання додаткових ознак, таких як Year та Month, позитивно вплинуло на результати.
3. Додаткові метрики: Для задачі з незбалансованими класами перевірте balanced accuracy або f1-score для більш точного аналізу.

              precision    recall  f1-score   support

           No       1.00      1.00      1.00     22672
          Yes       1.00      1.00      1.00      6420

    accuracy                            1.00     29092  
    macro avg       1.00      1.00      1.00     29092
    weighted avg    1.00      1.00      1.00     29092
